File: ui/BankingUiCommon/src/commonMain/kotlin/net/dankito/banking/ui/util/FlickerCodeAnimator.kt
package net.dankito.banking.ui.util

import kotlinx.coroutines.*
import net.dankito.banking.ui.model.tan.FlickerCode
import net.dankito.utils.multiplatform.log.LoggerFactory
import kotlin.jvm.Volatile


open class FlickerCodeAnimator {

    companion object {
        const val MinFrequency = 2
        const val MaxFrequency = 40
        const val DefaultFrequency = 20

        private val log = LoggerFactory.getLogger(FlickerCodeAnimator::class)
    }


    @Volatile
    protected var currentFrequency: Int = DefaultFrequency

    @Volatile
    protected var isPaused = false

    protected var animationJob: Job? = null



    open fun animateFlickerCode(flickerCode: FlickerCode, showStep: (Step) -> Unit) {
        stop() // stop may still running previous animation

        animationJob = GlobalScope.launch(Dispatchers.Default) {
            val steps = FlickerCodeStepsCalculator().calculateSteps(flickerCode.parsedDataSet)

            calculateAnimation(steps, showStep)
        }
    }

    protected open suspend fun calculateAnimation(steps: List<Step>, showStep: (Step) -> Unit) {
        var currentStepIndex = 0

        while (true) {
            if (isPaused == false) {
                val nextStep = steps[currentStepIndex]

                withContext(Dispatchers.Main) {
                    showStep(nextStep)
                }

                currentStepIndex++
                if (currentStepIndex >= steps.size) {
                    currentStepIndex = 0 // all steps shown, start again from beginning
                }
            }

            delay(1000L / currentFrequency)
        }
    }

    open fun pause() {
        this.isPaused = true
    }

    open fun resume() {
        this.isPaused = false
    }

    open fun stop() {
        try {
            animationJob?.cancel()

            animationJob = null
        } catch (e: Exception) {
            log.warn(e) { "Could not stop animation job" }
        }
    }


    open fun setFrequency(frequency: Int) {
        if (frequency in MinFrequency..MaxFrequency) {
            currentFrequency = frequency
        }
    }

}

File: ui/BankingiOSApp/BankingiOSApp/ui/views/tan/FlickerCodeAnimator.swift
import SwiftUI
import BankingUiSwift


class FlickerCodeAnimator {

    static let MinFrequency = 2
    static let MaxFrequency = 40
    static let DefaultFrequency = 30

    private var currentFrequency: Int = DefaultFrequency

    private var isPaused = false

    private var currentStepIndex = 0
    
    private var steps: [Step] = []
    
    private var showStep: ((Step) -> Void)? = nil

    private var timer: Timer? = nil
    


    func animate(_ flickerCode: String, _ showStep: @escaping (Step) -> Void) {
        self.stop() // stop may still running previous animation
        
        self.steps = FlickerCodeStepsCalculator().calculateSteps(flickerCode: flickerCode)
        
        self.showStep = showStep
                
        self.start()
    }
    
    
    private func start() {
        currentStepIndex = 0
        
        let interval = 1.0 / Double(currentFrequency)
        
        timer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { timer in
            self.calculateAndShowNextStep()
        }
    }

    private func calculateAndShowNextStep() {
        if isPaused == false {
            let nextStep = steps[currentStepIndex]
            
            self.showStep?(nextStep)

            currentStepIndex = currentStepIndex + 1
            
            if (currentStepIndex >= steps.count) {
                currentStepIndex = 0 // all steps shown, start again from beginning
            }
        }
    }

    func stop() {
        timer?.invalidate()
        
        timer = nil
    }

    func pause() {
        self.isPaused = true
    }

    func resume() {
        self.isPaused = false
    }


    func setFrequency(frequency: Int) {
        if frequency >= Self.MinFrequency && frequency <= Self.MaxFrequency {
            let isRunning = timer != nil
            
            stop()
            
            currentFrequency = frequency
            
            if isRunning {
                start()
            }
        }
    }
    
}
